### PR TITLE
Beauty

### PR DESCRIPTION
**Pull request by easyfrog**
Function could preview its parameters and return type directly in completion dialog.
ternjs4

### CODE REVIEW

File: tern.py [hunk 1]
```diff
@@ -421,14 +428,22 @@ def ensure_completions_cached(pfile, view):
...
-    if arg_completion_enabled and completion_icon(rec_type) == " (fn)":
+    if arg_completion_enabled and rec_type is not None and rec_type.startswith("fn("):
+      retval = parse_function_type(rec).get('retval')
+
+      if retval is None or retval == "()":
+        retval = ""
+      elif retval.startswith("{"):
+        retval = "{}"
+      elif retval.startswith("["):
+        retval = "[]"
```
Comment (marijnh) [inline tern.py]: Why are you removing information from the return type?

Comment (easyfrog) [inline tern.py]: I think the user most of the time just want to know the type of return value, and it looks more clean.

File: tern.py [hunk 2]
```diff
@@ -264,7 +264,9 @@ def f(port, doc):
       req = opener.open("http://" + localhost + ":" + str(port) + "/", json.dumps(doc).encode("utf-8"), 1)
       return json.loads(req.read().decode("utf-8"))
     except urllib.error.URLError as error:
-      raise Req_Error((hasattr(error, "read") and error.read().decode("utf-8")) or error.reason)
+      # error on mac
+      # raise Req_Error((hasattr(error, "read") and error.read().decode("utf-8")) or error.reason)
+      raise Req_Error(error.read().decode("utf-8"))
```
Comment (easyfrog) [inline tern.py]: OK . if use this codes: 
`raise Req_Error((hasattr(error, "read") and error.read().decode("utf-8")) or error.reason)`

and make sublime into background (switch to other app) and wait a while then go back to sublime. and type some code to invoke completion then will throw and error below:

```reloading plugin tern_for_sublime.tern
Traceback (most recent call last):
  File "./python3.3/urllib/request.py", line 1248, in do_open
  File "./python3.3/http/client.py", line 1065, in request
  File "./python3.3/http/client.py", line 1103, in _send_request
  File "./python3.3/http/client.py", line 1061, in endheaders
  File "./python3.3/http/client.py", line 906, in _send_output
  File "./python3.3/http/client.py", line 844, in send
  File "./python3.3/http/client.py", line 822, in connect
  File "./python3.3/socket.py", line 435, in create_connection
  File "./python3.3/socket.py", line 426, in create_connection
ConnectionRefusedError: [Errno 61] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/fengmap/Library/Application Support/Sublime Text 3/Packages/tern_for_sublime/tern.py", line 264, in f
    req = opener.open("http://" + localhost + ":" + str(port) + "/", json.dumps(doc).encode("utf-8"), 1)
  File "./python3.3/urllib/request.py", line 469, in open
  File "./python3.3/urllib/request.py", line 487, in _open
  File "./python3.3/urllib/request.py", line 447, in _call_chain
  File "./python3.3/urllib/request.py", line 1274, in http_open
  File "./python3.3/urllib/request.py", line 1251, in do_open
urllib.error.URLError: <urlopen error [Errno 61] Connection refused>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/fengmap/Library/Application Support/Sublime Text 3/Packages/tern_for_sublime/tern.py", line 322, in run_command
    data = make_request(port, doc)
  File "/Users/fengmap/Library/Application Support/Sublime Text 3/Packages/tern_for_sublime/tern.py", line 268, in f
    raise Req_Error((hasattr(error, "read") and error.read().decode("utf-8")) or error.reason)
tern_for_sublime.tern.Req_Error: <unprintable Req_Error object>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Applications/Sublime Text.app/Contents/MacOS/sublime_plugin.py", line 591, in on_query_completions
    res = callback.on_query_completions(v, prefix, locations)
  File "/Users/fengmap/Library/Application Support/Sublime Text 3/Packages/tern_for_sublime/tern.py", line 64, in on_query_completions
    completions, fresh = ensure_completions_cached(pfile, view)
  File "/Users/fengmap/Library/Application Support/Sublime Text 3/Packages/tern_for_sublime/tern.py", line 423, in ensure_completions_cached
    data = run_command(view, {"type": "completions", "types": True, "includeKeywords": True})
  File "/Users/fengmap/Library/Application Support/Sublime Text 3/Packages/tern_for_sublime/tern.py", line 324, in run_command
    if not silent: report_error(str(e), pfile.project)
TypeError: __str__ returned non-string (type ConnectionRefusedError)```

Comment (easyfrog) [inline tern.py]: OK . if use this codes: 
`raise Req_Error((hasattr(error, "read") and error.read().decode("utf-8")) or error.reason)`

and make sublime into background (switch to other app) and wait a while then go back to sublime. and type some code to invoke completion then will throw and error below:

```reloading plugin tern_for_sublime.tern
Traceback (most recent call last):
  File "./python3.3/urllib/request.py", line 1248, in do_open
  File "./python3.3/http/client.py", line 1065, in request
  File "./python3.3/http/client.py", line 1103, in _send_request
  File "./python3.3/http/client.py", line 1061, in endheaders
  File "./python3.3/http/client.py", line 906, in _send_output
  File "./python3.3/http/client.py", line 844, in send
  File "./python3.3/http/client.py", line 822, in connect
  File "./python3.3/socket.py", line 435, in create_connection
  File "./python3.3/socket.py", line 426, in create_connection
ConnectionRefusedError: [Errno 61] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/fengmap/Library/Application Support/Sublime Text 3/Packages/tern_for_sublime/tern.py", line 264, in f
    req = opener.open("http://" + localhost + ":" + str(port) + "/", json.dumps(doc).encode("utf-8"), 1)
  File "./python3.3/urllib/request.py", line 469, in open
  File "./python3.3/urllib/request.py", line 487, in _open
  File "./python3.3/urllib/request.py", line 447, in _call_chain
  File "./python3.3/urllib/request.py", line 1274, in http_open
  File "./python3.3/urllib/request.py", line 1251, in do_open
urllib.error.URLError: <urlopen error [Errno 61] Connection refused>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/fengmap/Library/Application Support/Sublime Text 3/Packages/tern_for_sublime/tern.py", line 322, in run_command
    data = make_request(port, doc)
  File "/Users/fengmap/Library/Application Support/Sublime Text 3/Packages/tern_for_sublime/tern.py", line 268, in f
    raise Req_Error((hasattr(error, "read") and error.read().decode("utf-8")) or error.reason)
tern_for_sublime.tern.Req_Error: <unprintable Req_Error object>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Applications/Sublime Text.app/Contents/MacOS/sublime_plugin.py", line 591, in on_query_completions
    res = callback.on_query_completions(v, prefix, locations)
  File "/Users/fengmap/Library/Application Support/Sublime Text 3/Packages/tern_for_sublime/tern.py", line 64, in on_query_completions
    completions, fresh = ensure_completions_cached(pfile, view)
  File "/Users/fengmap/Library/Application Support/Sublime Text 3/Packages/tern_for_sublime/tern.py", line 423, in ensure_completions_cached
    data = run_command(view, {"type": "completions", "types": True, "includeKeywords": True})
  File "/Users/fengmap/Library/Application Support/Sublime Text 3/Packages/tern_for_sublime/tern.py", line 324, in run_command
    if not silent: report_error(str(e), pfile.project)
TypeError: __str__ returned non-string (type ConnectionRefusedError)```

**Comment by marijnh** [comment]
Merged as 44f006c

**Pull request closed without merging**